apps/indexer/lib/indexer/fetcher/celo_epoch_rewards.ex

```diff
@@ -0,0 +1,118 @@
+defmodule Indexer.Fetcher.CeloEpochRewards do
+  @moduledoc """
+  Fetches Celo voter rewards for groups.
+  """
+  use Indexer.Fetcher
+  use Spandex.Decorators
+
+  require Logger
+
+  alias Ecto.Multi
+
+  alias Explorer.Celo.AccountReader
+  alias Explorer.Chain
+  alias Explorer.Chain.CeloEpochRewards
+
+  alias Indexer.BufferedTask
+  alias Indexer.Fetcher.Util
+
+  @behaviour BufferedTask
+
+  @doc false
+  def child_spec([init_options, gen_server_options]) do
+    Util.default_child_spec(init_options, gen_server_options, __MODULE__)
+  end
+
+  @impl BufferedTask
+  def init(initial, reducer, _json_rpc_named_arguments) do
+    {:ok, final} =
+      Chain.stream_blocks_with_unfetched_epoch_rewards(initial, fn block, acc ->
+        block
+        |> reducer.(acc)
+      end)
+
+    final
+  end
+
+  @impl BufferedTask
+  def run(entries, _json_rpc_named_arguments) do
+    failed_list =
+      entries
+      |> fetch_from_blockchain()
+      |> import_items()
+
+    if failed_list == [] do
+      :ok
+    else
+      {:retry, failed_list}
+    end
+  end
+
+  def fetch_from_blockchain(blocks) do
+    blocks
+    |> Enum.map(fn block ->
+      case AccountReader.validator_group_reward_data(block) do
+        {:ok, data} ->
+          data
+
+        error ->
+          Logger.debug(inspect(error))
+          Map.put(block, :error, error)
+      end
+    end)
+  end
+
+  def import_items(rewards) do
+    {failed, success} =
+      Enum.reduce(rewards, {[], []}, fn
+        %{error: _error} = reward, {failed, success} ->
+          {[reward | failed], success}
+
+        reward, {failed, success} ->
+          changeset = CeloEpochRewards.changeset(%CeloEpochRewards{}, reward)
+
+          if changeset.valid? do
+            {failed, [changeset.changes | success]}
+          else
+            {[reward | failed], success}
+          end
+      end)
+
+    import_params = %{
+      celo_epoch_rewards: %{params: success},
+      timeout: :infinity
+    }
+
+    if failed != [] do
+      Logger.debug(fn -> "requeuing rewards" end,
+        block_numbers: Enum.map(failed, fn rew -> rew.block_number end)
+      )
+    end
+
+    result =
+      Multi.new()
+      |> Multi.run(:import_rewards, fn _, _ ->
+        result = Chain.import(import_params)
+        {:ok, result}
+      end)
+      |> Multi.run(:delete_celo_pending, fn _, _ ->
+        success
+        |> Enum.each(fn reward -> Chain.delete_celo_pending_epoch_operation(reward.block_hash) end)
+
+        {:ok, success}
+      end)
+      |> Explorer.Repo.transaction()
```

Comment:
Normally the `Indexer` stays away from accessing the db directly, db transactions and import formatting is handled by the `Explorer` lib. See

* https://github.com/celo-org/blockscout/blob/57ded749ae41a71f1963cf186a0d5ffb1b1674f7/apps/indexer/lib/indexer/fetcher/internal_transaction.ex#L286
* https://github.com/celo-org/blockscout/blob/57ded749ae41a71f1963cf186a0d5ffb1b1674f7/apps/indexer/lib/indexer/block/catchup/fetcher.ex#L142
* https://github.com/celo-org/blockscout/blob/57ded749ae41a71f1963cf186a0d5ffb1b1674f7/apps/indexer/lib/indexer/fetcher/token.ex#L79

In these examples the importing of data to the db, and exact manner in which it's done is managed by a method in the `Explorer.Chain` context. This separates the concerns of fetching + processing data (`Indexer`), and verification + db import (`Explorer`).

tl;dr - this method needs to be moved into the `Explorer` app to avoid database logic + Ecto imports in the `Indexer`

Reply:
I thought about this and then saw that Ecto is indeed imported in some places in the `Indexer`, but you're right we shouldn't exacerbate this